### PR TITLE
Added error handling to dirEntries

### PR DESCRIPTION
**Pull request by marler8997**
Idea to fix https://issues.dlang.org/show_bug.cgi?id=12513.

This PR contains a proposed API change to `dirEntires` that allows the user to pass a callback delegate to handle errors.  This PR is for demonstration purposes only to show how the idea could work.  With the current PR, you should be able to iterate through a set of files that the program does not have permission to access, and the error handler callback should be called.

Here is an example program to demonstrate:

```D
import std.stdio;
import std.file;

struct MyErrorHandlers
{
    bool fileError(FileError error, const(char)[] name) nothrow @safe {
        writefln("[%s] %s", error, name);
        return false;
    }
}
void main()
{
    MyErrorHandlers errorHandlers;
    foreach(entry; dirEntries(getcwd(), SpanMode.breadth, true, &errorHandlers.fileError)) {
        writeln(entry.name);
    }
}
```

### CODE REVIEW

**Comment by DmitryOlshansky** [comment]
> I'll take a look into what it would take to redesign DirIteratorImpl to support recovering from an exception.

This is an important piece of work. @marler8997 are you around to play with it or should I close this pull?

**Comment by marler8997** [comment]
It's been a while since i looked at this. I'll take another look and get back to you soon.

**Comment by JakobOvrum** [comment]
Is this obsoleted by a combination of std.exception's `ifThrown` and `handle`?

**Comment by marler8997** [comment]
I'm not familiar with the 'handle' method, but if it allows you to handle exceptions and continue on with a range then it could be argued that it could replace this PR.  Have you tested it?

**Comment by marler8997** [comment]
I've looked at the handle template and I don't think DirIterator would be able to recover handling the exception.  I haven't looked at the implementation of handle but my guess is that it catches the exception and then continues using it as if nothing happens.  The problem with DirIterator is once the exception is thrown, the internal data structures have been corrupted.  In order to add support for the range to recover from throwing an exception, the range will likely need more memory to store the lost context after throwing the exception.

So to summarize, I don't think we have a solution quite yet.  It's would be nice to have a common solution to this problem.  If we can rework DirIterator to be exception safe without too much overhead, maybe that's the best solution.  It's hard to say at this point.  Sorry this wasn't much help, I've provided more questions than answers.

**Comment by quickfur** [comment]
ping
It's been a while. Any progress on this, or should we close it?

**Comment by marler8997** [comment]
We can close this PR, but I would make sure there is either an issue documented, or there is already a solution that allows you to iterate over directories and handle errors...is there a way to do that now?

**Comment by quickfur** [comment]
Is https://issues.dlang.org/show_bug.cgi?id=12513 good enough for this purpose? If not, please file a new issue (possibly referencing 12513 just so we don't lose track of it), and we can close this PR.

Thanks!

**Comment by marler8997** [comment]
That issue works.
